File: Uebung2BINT20/app/src/main/java/com/dasgrau/uebung2/MainActivity.java
package com.dasgrau.uebung2;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {

    private int counter = 0;
    private TextView textView;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        textView = findViewById(R.id.textView);

        findViewById(R.id.button).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                counter++;
                textView.setText(getString(R.string.hello_count, counter));
            }
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        textView.setText(getString(R.string.hello_count, counter));
    }

    @Override
    protected void onRestoreInstanceState(@NonNull Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        counter = savedInstanceState.getInt("counter", 0);
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        outState.putInt("counter", counter);
        super.onSaveInstanceState(outState);
    }
}

File: Uebung6_1BINT20/app/src/main/java/com/dasgrau/uebung6_1/MainActivity.kt
package com.dasgrau.uebung6_1

import android.Manifest
import android.content.pm.PackageManager
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.provider.CallLog
import android.util.Log
import android.widget.ListView
import androidx.annotation.RequiresPermission
import androidx.cursoradapter.widget.SimpleCursorAdapter

class MainActivity : AppCompatActivity() {

    companion object {
        val TAG = MainActivity::class.toString()
        const val REQUEST_CODE = 42
        val CALL_LOG_COLUMNS = arrayOf(CallLog.Calls._ID, CallLog.Calls.NUMBER,
            CallLog.Calls.DURATION)

        val ADAPTER_FROM = arrayOf(CallLog.Calls.NUMBER, CallLog.Calls.DURATION)
        val ADAPTER_TO = intArrayOf(R.id.textViewNumber, R.id.textViewDuration)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

       if(checkSelfPermission(Manifest.permission.READ_CALL_LOG)
           != PackageManager.PERMISSION_GRANTED) {
           Log.i(TAG, "Permission not granted, requesting...")
           requestPermissions(arrayOf(Manifest.permission.READ_CALL_LOG), REQUEST_CODE)
       } else {
           Log.i(TAG, "Permission already granted")
           queryCalls()
       }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        Log.i(TAG, "onRequestPermissionsResult with requestCode $requestCode" +
                " and grantResults[0] $grantResults[0]")
        if(requestCode != REQUEST_CODE || grantResults[0] != PackageManager.PERMISSION_GRANTED) {
            Log.i(TAG, "...permission NOT granted")
            return
        }
        Log.i(TAG, "...permission granted")
        queryCalls()
    }

    private fun queryCalls() {
        var cursor = contentResolver.query(CallLog.Calls.CONTENT_URI, CALL_LOG_COLUMNS,
            null, null, "${CallLog.Calls.LAST_MODIFIED} DESC")

        var adapter = SimpleCursorAdapter(applicationContext, R.layout.call_card, cursor,
            ADAPTER_FROM, ADAPTER_TO, 0)

        findViewById<ListView>(R.id.listViewCalls).adapter = adapter
    }
}